api/agent/protocol/json.go

```diff
@@ -0,0 +1,114 @@
+package protocol
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+)
+
+// This is sent into the function
+// All HTTP request headers should be set in env
+type JSONIO struct {
+	Headers    http.Header `json:"headers,omitempty"`
+	Body       string      `json:"body"`
+	StatusCode int         `json:"status_code,omitempty"`
+}
+
+// JSONProtocol converts stdin/stdout streams from HTTP into JSON format.
+type JSONProtocol struct {
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *JSONProtocol) IsStreamable() bool {
+	return true
+}
+
+type RequestEncoder struct {
+	*json.Encoder
+}
+
+func (h *JSONProtocol) DumpJSON(w io.Writer, req *http.Request) error {
+	stdin := json.NewEncoder(h.in)
+	_, err := io.WriteString(h.in, `{`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+
+	bb := new(bytes.Buffer)
+	_, err = bb.ReadFrom(req.Body)
+	if err != nil {
+		return err
+	}
+	reqData := bb.String()
+	if reqData != "" {
+		_, err := io.WriteString(h.in, `"body": `)
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		err = stdin.Encode(reqData)
+		if err != nil {
+			return err
+		}
+		_, err = io.WriteString(h.in, `,`)
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		defer bb.Reset()
+	}
+	_, err = io.WriteString(h.in, `"headers:"`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	err = stdin.Encode(req.Header)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	_, err = io.WriteString(h.in, `"}`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	return nil
+}
```

Comment:
the 'this shouldn't happen' comments can all be removed, as well, they just take up space. if they shouldn't happen then don't check them. 

in particular, encoding the request body can throw an error from encoding and other reasons, and the headers could maybe throw an error since we're injecting some of them. reading from the request body can also timeout (i/o read timeout on the conn) and the pipe for `h.in` could also be closed, so that could fail. 

a somewhat nice pattern to mitigate error handling is to do something like:

```go
func writeString(err error, dst io.Writer, str string) error {
  if err != nil {
     return err
  }
  _, err = io.WriteString(dst, str)
  return err
}

func Dispatch(...) error {
// ...
err = writeString(err, h.in, "{")
err = writeString(err, h.in, `"body":`)
// ...
return err
```

this writeString func is small enough for the compiler to inline it. a function closure is nicer looking but induces a function allocation, so it's better not to do that for this particular piece of code. anyway, at least remove the comments, but it most definitely _can_ happen and we do need to try to catch the error if we can, even though we could probably try to continue writing to a broken pipe and just return the last error, it's not foolproof